sycl/test-e2e/README.md

```diff
@@ -10,19 +11,95 @@
 
 # Overview
 This directory contains SYCL-related tests distributed in subdirectories based
-on testing scope.
+on testing scope. They follow the same style as other LIT tests but have a minor
+but crucial difference in the behavior of certain directives.
+
+First, some background. SYCL end-to-end tests take much longer time to compile
+than they spend on execution on a device. As such, we want to be able to
+structure the test in such a way that we compile it once and then execute
+multiple times on different devices (via `ONEAPI_DEVICE_SELECTOR`) to get
+required test coverage. The issue here is that the standard approach to `RUN`
+directives and substitutions doesn't allow us to expand a line into only
+dynamically known number of commands.
+
+To overcome that, these SYCL end-to-end tests implement a subclass of
+`lit.formats.ShTest` that adjusts default behavior to enable our usage scenario.
+Mainly, we introduce `%{run}` *expansion* that generates multiple commands from
+a single `RUN` directive - one per device in `sycl_devices` parameter (more
+on the parameter below). There is a small number of tests that either need
+multiple devices or target the device selector itself. For such tests we have a
+regular `%{run-unfiltered-devices}` substitution that doesn't set a
+`ONEAPI_DEVICE_SELECTOR` filter nor does it create multiple commands.
+
+This custom LIT test format also overrides the meaning of
+`REQUIRES`/`UNSUPPORTED`/`XFAIL` directives, although in a natural way that
+suites the `%{run}` expansion described above. First, "features" are split into
+device independent (e.g. "linux" or "cuda_dev_kit") and device dependent
+("cpu/gpu/accelerator", "opencl/cuda/hip/level_zero" and multiple "aspect-\*").
+Second, for each device in `sycl_devices` LIT parameter, we see if it satisfies
+the conditions in `UNSUPPORTED`/`REQUIRES` rules. If none of the devices do,
+the entire test is skipped as `UNSUPPORTED`. Otherwise, if multiple such devices
+are supported we do an additional filtering treating any `XFAIL` directives same
+way as `UNSUPPORTED`. If only one device is matched by `UNSUPPORTED`/`REQUIRES`
+filtering, then `XFAIL` behaves as an actual "expected to fail" marking.
+
+For any device left after the filtering above we expand each `RUN` directive
```

Comment:
Could you clarify if multi-line directives are supported, please?

Reply:
They are and we have such instances in our tests. That said, I wouldn't be very surprised if it's possible to construct a weird one that would break it, although I'm not sure how exactly one could do that.